In [ ]:
import pandas as pd
import numpy as np

# <mark style="background: #801010; color: #ffffff;" ></mark> NumPy Arrays


Create a python program that has a 1-D array with 5 elements. It should contain the numbers 0 to 4 in sequence. Assign this array to `output1`.

In [13]:
import numpy as np

# Write your solution here


output1 = np.arange(5)
print(output1)


[0 1 2 3 4]



Create a python program that contains a 2-D array with 6 elements (2 rows and 3 columns) that represents the following array. Assign this array to `output2`.

$$\begin{bmatrix} 0 & 1 & 2 \\ 3 & 4 & 5 \end{bmatrix}$$

In [11]:
import numpy as np

# Write your solution here
#output2 = np.array([[0,1,2], [3,4,5]])
#print(output2)
#rows, columns = output2.shape
#print(rows,columns)
output2 = np.arange(6).reshape(2,3)
print(output2)

[[0 1 2]
 [3 4 5]]


Create a python program that has a 2-D array with 6 elements (3 rows and 2 columns) that represents the following array. Assign this array to `output3`.

$$\begin{bmatrix} 0 & 1 \\ 2 & 3 \\ 4 & 5 \end{bmatrix}$$

In [12]:
import numpy as np

# Write your solution here
output3 = np.arange(6).reshape(3,2)
print(output3)


[[0 1]
 [2 3]
 [4 5]]



Create a python program that has a 3-D array with 8 elements (2 rows and 2 columns, with a depth of 2) using the `zeros` function. Assign this array to `output4`.

In [20]:
import numpy as np

# Write your solution here
output4 = np.zeros((2, 2, 2))
print(output4)

[[[0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]]]



You are given a 2-D array of unknown size named `array5`. Create a python program that finds the size of that array and returns the number of rows.

In [163]:
import numpy as np

def question05(array5):
    size = array5.shape
    rows, columns = array5.shape
    return rows
question05(np.arange(9).reshape(3,3))


3


Create a python program that performs the following calculation using universal functions. Assign the result to `output6`.

$$\begin{bmatrix} 0 & 1 \\ 2 & 3 \end{bmatrix} + \begin{bmatrix} 0 & 1 \\ 2 & 3 \end{bmatrix}$$

In [160]:
import numpy as np

# Write your solution here
output6_1 = np.array([[0,1],[2,3]])
output6_2 = np.array([[0,1],[2,3]])
output6 = np.add(output6_1, output6_2)
print(output6)

[[0 2]
 [4 6]]


You are given two 2-D arrays of that are the same size (`array7_i` and `array7_ii`). Create a python program that multiplies the two arrays together using universal functions. Return a new array containing the result.

In [ ]:
import numpy as np

def question07(array7_i,array7_ii):
    x = np.multiply(array7_i, array7_ii)
    return x
    # Write your solution here

question07(array7_i, array7_ii)

NameError: name 'array7_i' is not defined

Create a python program that performs the following calculation using universal functions. Assign the result to `output8`.

$$\begin{bmatrix} 0 & 1 \\ 2 & 3 \end{bmatrix} ** 3$$

In [159]:
import numpy as np

# Write your solution here
output8_i = np.array([[0,1],[2,3]])
output8 = np.power(output8_i, 3)
print(output8)

[[0 1]
 [2 3]]
[[ 0  1]
 [ 8 27]]


Create a python program that has a 1-D array with 100 elements. It should contain the numbers 0, 2, 4, 6, all the way to 198 in sequence. Assign this array to `output9`.

In [157]:
import numpy as np

# Write your solution here
output9 = np.arange(0,200,2)
print(output9)


[  0   2   4   6   8  10  12  14  16  18  20  22  24  26  28  30  32  34
  36  38  40  42  44  46  48  50  52  54  56  58  60  62  64  66  68  70
  72  74  76  78  80  82  84  86  88  90  92  94  96  98 100 102 104 106
 108 110 112 114 116 118 120 122 124 126 128 130 132 134 136 138 140 142
 144 146 148 150 152 154 156 158 160 162 164 166 168 170 172 174 176 178
 180 182 184 186 188 190 192 194 196 198]


Create a python program that has a 1-D array with 16 elements. It should contain the numbers 10, 20, 30, 40, all the way to 160 in sequence. Afterwards, change its shape using `reshape` to a 4-D array. Assign this reshaped array to `output10`.

***
<details>
  <summary><span style="color:blue">Click to see Hints</span></summary>
   In one of the provided examples, you saw how a 2-D 3-by-3 array. Use that as a basis for reshaping into a 4-D 2-by-2-by-2-by-2 array. 
</details>

In [192]:
import numpy as np

# Write your solution here
output10 = np.arange(10,161,10)
print(output10)
output10 = output10.reshape(2,2,2,2)
print(output10)

[ 10  20  30  40  50  60  70  80  90 100 110 120 130 140 150 160]
[[[[ 10  20]
   [ 30  40]]

  [[ 50  60]
   [ 70  80]]]


 [[[ 90 100]
   [110 120]]

  [[130 140]
   [150 160]]]]



One of the most common models in statistics (and by extention data science) is the normal distribution. In order to graph a normal distribution first the mean and standard deviation of the data must be found. NumPy has functions that allow us to analyse data and obtain these values.

You are given an array of unknown size and dimensions named `array11` that represents some data. Create a python program that finds the both the mean and standard deviation (rounded to 4 decimal points) of the data. Return the values in a dictionary of the following structure:
```python
{
  "mean": some_value,
  "standard_deviation": some_value
}
```

In [191]:
import numpy as np

def question11(array11):
    mean = np.mean(array11)
    standard_deviation = round(np.std(array11),4)
    dict = {"mean": mean , "standard_deviation": standard_deviation}
    return dict
    # Write your solution here

question11(np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4, 5, 6, 7, 8, 9]))['standard_deviation']


2.582


You are given an 1-D array of unknown size named `array12`. Create a python program that finds both the first element and the smallest element in the array. Return the values in a dictionary of the following structure:
```python
{
  "first": some_value,
  "smallest": some_value
}
```

In [63]:
import numpy as np

def question12(array12):
    first_ele = array12[0]
    smallest_ele = np.min(array12)
    dict = {"first": first_ele , "smallest": smallest_ele}
    return dict
    # Write your solution here
question12(np.arange(100,201,2))

{'first': 100, 'smallest': 100}


You are given an 2-D array of unknown size named `array13`. Create a python program that finds the first row of data and the maximum number within that row.  Return the values in a dictionary of the following structure:
```python
{
  "first_row": some_value,
  "max_row_value": some_value
}
```

In [244]:
import numpy as np

def question13(array13):
    first_row = array13[0:1:,0::][0]
    max_row = np.max(first_row)
    dict_ = {"first_row": first_row , "max_row_value": max_row}
    return dict_
   
question13(np.array([50,64,44,77,89,89,24,47,61,35,59,10,18,78,25,51,83,28,70,20,98,77,10,27,27,15,2,39,22,72,63,32,53,7,85,96,68,14,41,99,59,49,35,29,15,14,89,8,23,8,63,17,10,26,89,77,17,35,5,22,58,85,61,39,14,2,27,25,92,86,52,69,9,85,74,72,47,60,82,51,79,70,70,21,64,8,99,20,64,83,26,81,79,32,44,56,78,25,78,94]).reshape(25, 4))['first_row']

array([50, 64, 44, 77])


You are given an 1-D array of unknown size named `array14`. Create a python program that finds the first 6 entries within that row. Next, split that array into two arrays using the `split` function. Return the values in a dictionary of the following structure:
```python
{
  "array": first_6_entries,
  "part1": split_first_half,
  "part2": split_other_half
}
```

In [176]:
import numpy as np

def question14(array14):
    array = array14[0:6]
    part1, part2 = np.split(array,2)
    dict = {"array": array , "part1": part1, "part2":part2 }
    return dict
   
question14(np.arange(-10, 50, 5))['part2']


array([ 5, 10, 15])


You are given an 1-D array of 40 elements named `array15`. Create a python program that finds every second element starting from index 4 and stopping before index 20. Return the result.

In [172]:
import numpy as np

def question15(array15):
    z = array15[4:20:2 ]
    return z
   
question15(np.arange(-100, 100, 5))


array([-80, -70, -60, -50, -40, -30, -20, -10])


You are given 2 sorted 1-D arrays of 20 elements named `array16_i` and `array16_ii`. Create a python program that concatenates these two arrays by appending `array16_ii` to the end of `array16_i`. Afterwards, sort the new array.
Return both the unsorted concatenated array, and the sorted array in a dictionary of the following structure:
```python
{
  "concat": concatenated_array,
  "sorted": sorted_array
}
```

In [89]:
import numpy as np

def question16(array16_i, array16_ii):
    concat_array = np.concatenate([array16_i,array16_ii])
    sorted_array = np.sort(concat_array)
    dict = {"concat": concat_array , "sorted": sorted_array}
    return dict
   
question16(np.arange(100,104), np.arange(100,104))   

{'concat': array([100, 101, 102, 103, 100, 101, 102, 103]),
 'sorted': array([100, 100, 101, 101, 102, 102, 103, 103])}


Create a python program that:

1. initializes two arrays using the `arange` function followed by the `reshape` functions.
   1. The first array should have 2 rows and 5 columns and contain the numbers 0, 1, 2, up to and including 9.
   2. The second array should  have 2 rows and 3 columns and contain the numbers 1, 3, 5, up to and including 11.
2. concatenates these two arrays by appending along the column axis (creating an array with 2 rows and 8 columns), then transposes the result (creating an array with 8 rows and 2 columns).

Assign the end result to `output17`.

In [97]:
# Write your solution here
output17 = np.transpose(np.hstack([np.arange(10).reshape(2,5),np.arange(1,12,2).reshape(2,3)]))
print(output17)

[[ 0  5]
 [ 1  6]
 [ 2  7]
 [ 3  8]
 [ 4  9]
 [ 1  7]
 [ 3  9]
 [ 5 11]]



Create a python program that:

1. initializes two arrays using the `arange` function followed by the `reshape` functions.
   1. The first array should have 4 rows and 2 columns and contain the numbers 0, 1, 2, up to 7.
   2. The second array should have 3 rows and 2 columns and contain the numbers 0, 5, 10, up to 11.
2. concatenates these two arrays by appending along the row axis.
3. afterwards, reshape the new array into a 1-D array and sort it in decending order.

Assign the end result to `output18`.

In [252]:
# Write your solution here
output18 = -np.sort(-(np.vstack([np.arange(8).reshape(4,2),np.arange(1,12,2).reshape(3,2)])).reshape(1,14))
output18 = output18[0]
print(output18)

[11  9  7  7  6  5  5  4  3  3  2  1  1  0]



Create a python program that:

1. initializes two 2-D arrays as depicted below.
$$a1 = \begin{bmatrix} 70 & 51 & 89 \\ 42 & 15 & 8 \end{bmatrix}, a2 = \begin{bmatrix} 75 & 12 & 38 \\ 99 & 75 & 45 \end{bmatrix}$$
2. performs integer division, by dividing the elements of a1 by a2. Assign this result to `output19_i`.
3. sums the three arrays together. Assign this result to `output19_ii`, also assign a copy of the second row to `output19_iii`.


In [246]:
# Write your solution here
a1 = np.array([[70,51,89], [42,15,8]])
a2 = np.array([[75,12,38], [99,75,45]])
a1,a2
output19_i = np.floor_divide(a1,a2)
print(output19_i)
output19_ii = np.add(a1,a2,output19_i)
print(output19_ii)
output19_iii = output19_ii[1::].copy()[0]
print(output19_iii)

[[0 4 2]
 [0 0 0]]
[[145  63 127]
 [141  90  53]]
[141  90  53]



Create a python program that:

1. initializes two 2-D arrays as depicted below.
$$a1 = \begin{bmatrix} 48 & 75 & 75 \\ 15 & 2 & 46 \end{bmatrix}, a2 = \begin{bmatrix} 78 & 15 & 87 \\ 0 & 61 & 9 \end{bmatrix}$$
2. multiplies the two arrays together, then slice the result to only assign the last column to `output20` (note: an n-row, 1-column array automatically converts into a 1-row, n-column array, leave your answer like this).


In [245]:
# Write your solution here
a1 = np.array([[48,75,75], [15,2,46]])
a2 = np.array([[78,15,87], [0,61,9]])
a1,a2
output20 = np.multiply(a1,a2)
output20 = output20[0:2,2:3]
output20 = output20.reshape(1,2)[0]
print(output20)

[6525  414]
